### PR TITLE
WL 21551 - Flocking Fish Aquarium

### PR DESCRIPTION
**Pull request by cain-kilgore**
> replace the basic block with a nice looking model of a fish. 
> And when the script is run, also create a simple translucent walled 'aquarium' that the fish stay inside. 
> use a tablet button to make/delete the acquarium and fish. 
> put the results in the marketplace

Worklist [21551](https://worklist.net/21551)

## Test Plan
1. Run the FlockOfFish.js script from here: [FlockOfFish.js](https://s3-eu-west-1.amazonaws.com/cainkilgore-prod/files/FlockOfFish.js)
2. Confirm that a "FishTank" Toolbar/Tablet Button gets added
3. Confirm that clicking on it toggles the state off and on and a fishtank appears in front of you.
4. Make sure the fishes spawned in correctly and move around.
5. Confirm that when you leave a domain without stopping the script that the fish and Aquarium disappear.

Marketplace Item: [here (not available yet)](https://highfidelity.com/marketplace/items/4535b57c-f35a-4d9d-a3ee-1233c457dc8e
)

b5080c8c-5bf4-4d09-825e-63d2a2224792

### CODE REVIEW

**Comment by themelissabrown** [comment]
@liv once this has a code review and a sign off by MIlad, it can be merged and go to marketplace

File: Marketplace/aquarium/FlockOfFish.js [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+                        // Get only the properties we need, because that is faster
+                        var otherProps = flockProperties[j];
+                        var separation = Vec3.distance(properties.position, otherProps.position);
+                        if (separation < MAX_SIGHT_DISTANCE) {
+                            averageVelocity = Vec3.sum(averageVelocity, otherProps.velocity);
+                            averagePosition = Vec3.sum(averagePosition, otherProps.position);
+                            othersCounted++;
+                        }
+                        if (separation < MIN_SEPARATION) {
+                            var pushAway = Vec3.multiply(Vec3.normalize(Vec3.subtract(properties.position, otherProps.position)), AVOIDANCE_FORCE);
```
Comment (misslivirose) [inline Marketplace/aquarium/FlockOfFish.js]: Lines should be split when line length is greater than 128 characters

File: Marketplace/aquarium/FlockOfFish.js [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+                            velocity = Vec3.sum(velocity, pushAway);
+                        }
+                    }
+                }
+
+                if (othersCounted > 0) {
+                    averageVelocity = Vec3.multiply(averageVelocity, 1.0 / othersCounted);
+                    averagePosition = Vec3.multiply(averagePosition, 1.0 / othersCounted);
+                    //  Alignment: Follow group's direction and speed
+                    velocity = Vec3.mix(velocity, Vec3.multiply(Vec3.normalize(averageVelocity), Vec3.length(velocity)), ALIGNMENT_FORCE);
```
Comment (misslivirose) [inline Marketplace/aquarium/FlockOfFish.js]: Lines should be split when line length is greater than 128 characters

File: Marketplace/aquarium/FlockOfFish.js [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+                    // Cohesion: Steer towards center of flock
+                    var towardCenter = Vec3.subtract(averagePosition, position);
+                    velocity = Vec3.mix(velocity, Vec3.multiply(Vec3.normalize(towardCenter), Vec3.length(velocity)), COHESION_FORCE);
```
Comment (misslivirose) [inline Marketplace/aquarium/FlockOfFish.js]: Lines should be split when line length is greater than 128 characters

File: Marketplace/aquarium/FlockOfFish.js [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+                } 
+
+                //  Orient in direction of velocity 
+                var rotation = Quat.rotationBetween(Vec3.UNIT_NEG_Z, velocity);
+                var VELOCITY_FOLLOW_RATE = 0.30;
+
+                //  Only update properties if they have changed, to save bandwidth
+                var MIN_POSITION_CHANGE_FOR_UPDATE = 0.001;
+                if (Vec3.distance(properties.position, position) < MIN_POSITION_CHANGE_FOR_UPDATE) {
+                    Entities.editEntity(fish[i].entityId, { velocity: velocity, rotation: Quat.mix(properties.rotation, rotation, VELOCITY_FOLLOW_RATE) });
```
Comment (misslivirose) [inline Marketplace/aquarium/FlockOfFish.js]: Lines should be split when line length is greater than 128 characters

File: Marketplace/aquarium/FlockOfFish.js [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+                } else {
+                    Entities.editEntity(fish[i].entityId, { position: position, velocity: velocity, rotation: Quat.slerp(properties.rotation, rotation, VELOCITY_FOLLOW_RATE) });
```
Comment (misslivirose) [inline Marketplace/aquarium/FlockOfFish.js]: Lines should be split when line length is greater than 128 characters

File: Marketplace/aquarium/FlockOfFish.js [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+    
+    var STARTING_FRACTION = 0.25; 
+    function loadFish(howMany) {
+      var center = Vec3.sum(MyAvatar.position, Vec3.multiply(Quat.getFront(MyAvatar.orientation), 2 * TANK_WIDTH));
+      lowerCorner = { x: center.x - TANK_WIDTH / 2, y: center.y, z: center.z - TANK_WIDTH / 2 };
+      upperCorner = { x: center.x + TANK_WIDTH / 2, y: center.y + TANK_HEIGHT, z: center.z + TANK_WIDTH / 2 };
+      
+      for (var i = 0; i < howMany; i++) {
+        var position = { 
+            x: lowerCorner.x + (upperCorner.x - lowerCorner.x) / 2.0 + (Math.random() - 0.5) * (upperCorner.x - lowerCorner.x) * STARTING_FRACTION,
```
Comment (misslivirose) [inline Marketplace/aquarium/FlockOfFish.js]: line length

**Comment by Nex-Pro** [comment]
@cain-kilgore  It's nice to see you removed the cruft I told you about.

**Comment by Nex-Pro** [comment]
@themelissabrown @misslivirose @cain-kilgore Does this clean up if the avatar that created the aquarium leaves the domain or crashes ? Will the goldfish become Exocoetidae ?

**Comment by cain-kilgore** [comment]
@Nex-Pro They do indeed become flying fish. Do you have any suggestions as to the best way to prevent that from happening?

**Comment by MiladNazeri** [comment]
I had a chance to take a look and I noticed a few things with some of the calculations.
The way the random positions are being generated at first are oscillating between bounds outside of the walls, not a big thing, but it's why fish appear outside of the walls sometimes at first.
One other thing is that the fish can came outside of the aquarium because there isn't a margin. 

Taking a look at this gave me some ideas though, there are a lot of interesting calculations into the fish, and since it is already kind of a tablet app, instead of finding the perfect settings for how the fish work, giving some controls on how the fish tank looks and works can be interesting. 

I went through and did a quick version that automates some controls with an external JSON file and addresses some of the issues with the calculations:  
Might not be for this pull necessarily:

https://s3.us-east-2.amazonaws.com/hifi-fishtank/FlockOfFish.js
https://s3.us-east-2.amazonaws.com/hifi-fishtank/fishDefault.json
https://s3.us-east-2.amazonaws.com/hifi-fishtank/fishTankHTMLHandler.js
https://s3.us-east-2.amazonaws.com/hifi-fishtank/fishTankUI.html

* Just saw NexPros comment.  Putting a lifetime on the models might be a work around for the flying fish?

**Comment by cain-kilgore** [comment]
@misslivirose I've just reviewed your comments and made the changes. Let me know if this is what was wanted!

**Comment by MiladNazeri** [comment]
The fish look much better too as far as where they are first appearing and not poking their heads out of the aquarium!  

@misslivirose I'm not sure how picky you want to be with eslint rules, but I saw some eslint warnings for spacing errors and double =='s.

**Comment by misslivirose** [comment]
Thanks @MiladNazeri, I hadn't had a chance to review in an editor. @cain-kilgore - can you please make sure that this adheres to using double == and that the spacing in loadFish() is correct with indentation? It's currently 6 spaces instead of 8.

**Comment by cain-kilgore** [comment]
Sorry @misslivirose forgot to say changes were made. Yay Dev Approved!

**Comment by MiladNazeri** [comment]
Tested with Cain last night.  This doesn't leave fish around in a domain if the user leaves now.  

@themelissabrown I Recommend setting QA Approved and ready for merge and marketplace.

**Pull request merged**